_posts/2019-07-01-test-design.md

```diff
@@ -0,0 +1,296 @@
+
+In performing such a refactoring we should not need to change any test because we are not adding or changing any functionality, only moving logic around. 
+However, if we want to keep our *design* of having a test class for each production class, we need to refactor our tests as well.
+
+Even if we don't want to keep this design, our tests will have to be modified because chances are big that the API of our production code changed. 
+The parameters of methods might have changed, the fields of objects might have changed, constructors might have changed, etc.
+If we are lucky, the tests still compile, but they will very likely fail. 
+And the larger your application becomes, the more work it will be to get all tests compiling and green again.
+
+This phenomenon is known as [fragile tests](http://xunitpatterns.com/Fragile%20Test.html){:target="_blank" rel="noopener noreferrer"}.
+
+## Fragile tests example
+
+An example of this phenomenon that we encountered on a project is the creation of an instance of an [aggregate](https://martinfowler.com/bliki/DDD_Aggregate.html){:target="_blank" rel="noopener noreferrer"}.
+A lot of the tests in our project needed an instance of an aggregate. 
+This was not a problem at first, we just created aggregates by using the constructor of the class and passing all the necessary data in it.
+We created these instances in every test where we needed them, or sometimes created a method in the test class to not duplicate the construction too much in that class.
+
+To illustrate the issue we will look at a fictional simplified example about order creation.
+```java
+public class Order {
+    
+    private final UUID customerId;
+    
+    public Order(final Customer customer) {
+        Assert.notNull(customer, "Customer should not be null");
+        Assert.isTrue(customer.isActive(), "Customer should be active");
+        this.customerId = customer.getId();
+    }
+    
+    public UUID getCustomerId() {
+        return this.customerId;
+    }
+}
+```
+
+```java
+public class OrderTest {
+    
+    @Test
+    public void given_an_active_customer_then_order_creation_should_be_successful() {
+        final UUID customerId = UUID.randomUUID();
+        final Customer customer = mock(Customer.class);
+        when(customer.getId()).thenReturn(customerId);
+        when(customer.isActive()).thenReturn(true);
+        
+        final Order order = new Order(customer);
+        
+        assertThat(order.getCustomerId()).isEqualTo(customerId);
+    }
+    
+    @Test
+    public void given_an_inactive_customer_then_order_creation_should_result_in_an_illegal_argument_exception() {
+        final Customer customer = mock(Customer.class);
+        when(customer.isActive()).thenReturn(false);
+ 
+        assertThrows(IllegalArgumentException.class, () -> new Order(customer));
+    }
+}
+```
+
+Our problems began when we realised that the constructor of our aggregate was becoming too large. 
```

Comment:
I would rephrase here that the number of parameters became too large.
When I read the second part about the complex logic in the constructor, I thought you had written the same thing twice.